src/main/java/com/regnosys/rosetta/common/compile/JavaCSourceCancellableCompiler.java

```diff
@@ -0,0 +1,182 @@
+                                          Path... additionalClassPaths) {
+        this(DEFAULT_THREAD_POLL_INTERVAL_MS,
+                DEFAULT_MAX_COMPILE_TIMEOUT_MS,
+                executorService,
+                useSystemClassPath,
+                deleteOnError,
+                isVerbose,
+                releaseFlag,
+                additionalClassPaths);
+    }
+
+    @Override
+    public JavaCompilationResult compile(List<Path> sourceJavaPaths,
+                                         Path targetPath,
+                                         Supplier<Boolean> isCancelled) throws ExecutionException, InterruptedException, TimeoutException {
+        DiagnosticCollector<JavaFileObject> diagnosticCollector = new DiagnosticCollector<>();
+        JavaCompiler.CompilationTask compilationTask = getCompilationTask(sourceJavaPaths, targetPath, diagnosticCollector);
+
+        Future<Boolean> submittedTask = executorService.submit(compilationTask);
+
+        Optional<Boolean> result = submitAndWait(submittedTask, isCancelled);
+
+        if (deleteOnError && result.isPresent() && !result.get()) {
+            wipeTargetPath(targetPath);
+        }
+
+        return new JavaCompilationResult(result.isPresent(), result.orElse(false), diagnosticCollector.getDiagnostics());
+    }
+
+    private void wipeTargetPath(Path targetPath) {
+        try {
+            FileUtils.cleanDirectory(targetPath.toFile());
+        } catch (IOException e) {
+            throw new CompilationTargetDeletionException("Failed to delete target classes after compilation error", e);
+        }
+    }
+
+    private Optional<Boolean> submitAndWait(Future<Boolean> submittedTask, Supplier<Boolean> isCancelled) throws InterruptedException, ExecutionException, TimeoutException {
+        int maxWaitCycles = maxCompileTimeoutMs / threadPollIntervalMs;
+
+        for (int i = 0; i < maxWaitCycles; i++) {
+            try {
+                LOGGER.debug("Trying get from task");
+                return Optional.of(submittedTask.get(threadPollIntervalMs, TimeUnit.MILLISECONDS));
+            } catch (TimeoutException e) {
+                if (isCancelled.get()) {
+                    boolean cancellationAttempted = submittedTask.cancel(true);
+                    if (!cancellationAttempted && submittedTask.isCancelled()) {
+                        LOGGER.warn("Attempted to cancel a compilation task but the cancellation attempt was unsuccessful, this may be because the task was already cancelled");
+                    }
+                    if (!submittedTask.isCancelled()) {
+                        String message = "Attempt to cancel compilation task failed";
+                        LOGGER.error(message);
+                        throw new CompilationCancellationException(message);
+                    }
+                    return Optional.empty();
+                }
+                LOGGER.debug("Timed out whilst getting from task");
```

Comment:
Make trace + add the iteration number so its actually useful.

Reply:
Done